File: CQ_SDK/GroupMsgEvent.cpp
#include "GroupMsgEvent.h"

#include "API.h"
#include "code.h"

using namespace std;


CQ::GroupMsgEvent::GroupMsgEvent(int subType, int msgId, long long fromGroup, long long fromAccount, const char * fromAnonymous, const char * msg, int font)
	:MsgEvent(subType, msgId, fromAccount, msg, font), fromGroup(fromGroup), fromAnonymousToken(fromAnonymous), fromAnonymousInfo()
{}

CQ::GroupMsgEvent::~GroupMsgEvent()
{
	if (fromAnonymousInfo != nullptr)
		delete fromAnonymousInfo;
}

bool CQ::GroupMsgEvent::isAnonymous() const { return fromAccount == 80000000; }

AnonymousInfo & CQ::GroupMsgEvent::getFromAnonymousInfo() //throw(std::exception_ptr)
{
	if (isAnonymous())
		return
		fromAnonymousInfo != nullptr
		? *fromAnonymousInfo
		: *(fromAnonymousInfo = new AnonymousInfo(fromAnonymousToken));
	else
		throw std::exception_ptr();
}

bool CQ::GroupMsgEvent::setGroupKick(bool refusedAddAgain)
{
	return !CQ::setGroupKick(fromGroup, fromAccount, refusedAddAgain);
}

bool CQ::GroupMsgEvent::setGroupBan(long long banTime)
{
	if (isAnonymous()) {
		return !CQ::setGroupAnonymousBan(fromGroup, fromAnonymousToken, banTime);
	}
	else {
		return !CQ::setGroupBan(fromGroup, fromAccount, banTime);
	}
}

bool CQ::GroupMsgEvent::setGroupAdmin(bool isAdmin)
{
	return !CQ::setGroupAdmin(fromGroup, fromAccount, isAdmin);
}

bool CQ::GroupMsgEvent::setGroupSpecialTitle(std::string 专属头衔, long long 过期时间)
{
	return !CQ::setGroupSpecialTitle(fromGroup, fromAccount, 专属头衔, 过期时间);
}

bool CQ::GroupMsgEvent::setGroupWholeBan(bool 开启禁言)
{
	return CQ::setGroupWholeBan(fromGroup, 开启禁言);
}

bool CQ::GroupMsgEvent::setGroupAnonymous(bool 开启匿名)
{
	return CQ::setGroupAnonymous(fromGroup, 开启匿名);
}

bool CQ::GroupMsgEvent::setGroupCard(std::string 新名片_昵称)
{
	return CQ::setGroupCard(fromGroup, fromAccount, 新名片_昵称);
}

bool CQ::GroupMsgEvent::setGroupLeave(bool 是否解散)
{
	return CQ::setGroupLeave(fromGroup, 是否解散);
}

GroupMemberInfo CQ::GroupMsgEvent::getGroupMemberInfo(bool 不使用缓存)
{
	return CQ::getGroupMemberInfo(fromGroup, fromAccount, 不使用缓存);
}

std::vector<GroupMemberInfo> CQ::GroupMsgEvent::getGroupMemberList()
{
	return CQ::getGroupMemberList(fromGroup);
}
int CQ::GroupMsgEvent::sendMsg(const char * msg) const { return sendGroupMsg(fromGroup, msg); }
int CQ::GroupMsgEvent::sendMsg(std::string msg) const { return sendGroupMsg(fromGroup, msg); }
MsgSend CQ::GroupMsgEvent::sendMsg()const { return MsgSend(fromGroup, msgtype::群); }

int CQ::GroupMsgEvent::sendMsg(const char* msg, bool at_user) const { return sendMsg(string(msg), at_user); }
int CQ::GroupMsgEvent::sendMsg(std::string msg, bool at_user) const { return at_user ? sendMsg(code::at(fromAccount) + msg) : sendMsg(msg); }
MsgSend CQ::GroupMsgEvent::sendMsg(bool at_user) const { auto res = sendMsg(); if (at_user) res << code::at(fromAccount); return res; }

File: CQ_HELP/src/事件/GroupMsgEvent.java
/*
 * Copyright (C) 2018 admin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package 事件;

import c.string;
import c.vector;
import 信息.AnonymousInfo;
import 信息.GroupMemberInfo;

/**
 * 群聊事件
 * @author MikuPy2001
 */
public abstract class GroupMsgEvent extends MsgEvent {

    //群号
    public long fromGroup;
    //禁言用的令牌
    public string fromAnonymousToken;

    /**
     * 判断是否匿名
     *
     * @return 如果是匿名者发送的消息则返回真
     */
    public abstract boolean isAnonymous();

    /**
     * 获取匿名者信息
     *
     * @return 匿名者信息
     */
    public abstract AnonymousInfo getFromAnonymousInfo() /*throw(exception_ptr)*/;

    /**
     * 置群员移除
     *
     * @param refusedAddAgain 是否不再接收入群信息
     * @return 秘制返回
     * @see 酷Q.API#setGroupKick(long, long, boolean)
     */
    public abstract boolean setGroupKick(boolean refusedAddAgain);

    /**
     * 置群员禁言,自动判断是否是匿名.
     *
     * @param banTime 禁言时间,可省略,默认为60
     * @return 秘制返回
     * @see 酷Q.API#setGroupBan(long, long, long)
     * @see 酷Q.API#setGroupAnonymousBan(long, c.string, long)
     */
    public abstract boolean setGroupBan(long banTime);

    /**
     * 置群管理员
     *
     * @param isAdmin 可省略,默认为取消管理
     * @return 秘制返回
     * @see 酷Q.API#setGroupAdmin(long, long, boolean)
     */
    public abstract boolean setGroupAdmin(boolean isAdmin);

    /**
     * 置群成员专属头衔
     *
     * @param 专属头衔 专属头衔
     * @param 过期时间 单位(秒),可省略,默认永久(-1)
     * @return 秘制返回
     * @see 酷Q.API#setGroupSpecialTitle(long, long, c.string, long)
     */
    public abstract boolean setGroupSpecialTitle(string 专属头衔, long 过期时间);

    /**
     * 置全群禁言
     *
     * @param 开启禁言 可省略,默认为禁言,假为解禁
     * @return 秘制返回
     * @see 酷Q.API#setGroupWholeBan(long, boolean)
     */
    public abstract boolean setGroupWholeBan(boolean 开启禁言);

    /**
     * 置群匿名设置
     *
     * @param 开启匿名
     * @return 秘制返回
     * @see 酷Q.API#setGroupAnonymous(long, boolean)
     */
    public abstract boolean setGroupAnonymous(boolean 开启匿名);

    /**
     * 置群成员名片
     *
     * @param 新名片_昵称 新名片_昵称
     * @return 秘制返回
     * @see 酷Q.API#setGroupCard(long, long, c.string)
     */
    public abstract boolean setGroupCard(string 新名片_昵称);

    /**
     * 置群退出
     *
     * @param 是否解散 群主不可退出,退出既为解散群
     * @return 秘制返回
     * @see 酷Q.API#setGroupLeave(long, boolean)
     */
    public abstract boolean setGroupLeave(boolean 是否解散);

    /**
     * 取群成员信息 (默认使用缓存)
     *
     * @param 不使用缓存 强制获取最新信息,多次调用可能会冻结QQ,默认为假
     * @return 群成员信息
     * @see 酷Q.API#getGroupMemberInfo(long, long, boolean)
     */
    public abstract GroupMemberInfo getGroupMemberInfo(boolean 不使用缓存);

    /**
     * 取群成员列表
     *
     * @return 群成员列表
     * @see 酷Q.API#getGroupMemberList(long)
     */
    public abstract vector<GroupMemberInfo> getGroupMemberList();
}
